examples/fluidsim.dx

```diff
@@ -136,18 +108,40 @@ def fluidsim (dict: VSpace a) ?=> (num_steps: Int) (color_init: n=>m=>a)
 
 '### Demo
 
+N = Fin 50
+M = Fin 50
+
 -- Create random velocity field.
 def ixkey3 (k:Key) (i:n) (j:m) (k2:o) : Key =
   hash (hash (hash k (ordinal i)) (ordinal j)) (ordinal k2)
-v = for i:N j:M k:(Fin 2). 3.0 * (randn $ ixkey3 (newKey 0) i j k)
+init_velocity = for i:N j:M k:(Fin 2).
+  3.0 * (randn $ ixkey3 (newKey 0) i j k)
 
 -- Create diagonally-striped color pattern.
 init_color = for i:N j:M.
-  BToF $ (sin $ (IToF $ (ordinal j) + (ordinal i)) / 8.0) > 0.0
+  r = BToF $ (sin $ (IToF $ (ordinal j) + (ordinal i)) / 8.0) > 0.0
+  b = BToF $ (sin $ (IToF $ (ordinal j) - (ordinal i)) / 6.0) > 0.0
+  g = BToF $ (sin $ (IToF $ (ordinal j) + (ordinal i)) / 4.0) > 0.0
+  [r, g, b]
 
 -- Run fluid sim and plot it.
-num_steps = 50
-final_color = fluidsim num_steps init_color v
+num_steps = 5
+final_color = fluidsim num_steps init_color init_velocity
+
+:html imshow final_color
+> <html output>
+
+
+
+'### Gradient test
+
+target = transpose init_color
+
+def objective (v:N=>M=>(Fin 2)=>Float) : Float =
+  final_color = fluidsim num_steps init_color v
+  sum for (i, j, c). sq (final_color.i.j.c - target.i.j.c)
+
+init_vel_grad = grad objective zero
 
-:html matshow final_color
+:html imshow for i j. [0.0, init_vel_grad.i.j.(0@_), init_vel_grad.i.j.(1@_)]
```

Comment:
I wonder what is the significance of the colors in the image for the primal fluid simulation computation and the gradient computation?

Reply:
I'm guessing that more color means greater numerical magnitude, less color means closer to zero.